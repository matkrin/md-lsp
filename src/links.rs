use lsp_types::Url;
use markdown::{
    mdast::{Heading, Link, Node, Text},
    unist::{Point, Position as AstPosition},
};
use regex::Regex;

use crate::{
    ast::{get_heading_text, TraverseNode},
    state::State,
};

use percent_encoding::{AsciiSet, CONTROLS};

/// https://url.spec.whatwg.org/#fragment-percent-encode-set
const FRAGMENT: &AsciiSet = &CONTROLS.add(b' ').add(b'"');

pub fn url_encode(s: &str) -> String {
    percent_encoding::percent_encode(s.as_bytes(), FRAGMENT).to_string()
}

pub fn url_decode(s: &str) -> String {
    percent_encoding::percent_decode(s.as_bytes())
        .decode_utf8_lossy()
        .to_string()
}

#[derive(Debug)]
pub enum MdLink<'a> {
    NormalLink(&'a Link),
    WikiLink(&'a Link),
}

impl<'a> MdLink<'a> {
    fn new(link: &'a Link) -> Self {
        match &link.title {
            Some(title) if title == "wikilink" => MdLink::WikiLink(link),
            _ => MdLink::NormalLink(link),
        }
    }

    pub fn position(&self) -> Option<&AstPosition> {
        match self {
            MdLink::NormalLink(link) | MdLink::WikiLink(link) => link.position.as_ref(),
        }
    }

    pub fn title(&self) -> Option<&str> {
        match self {
            MdLink::NormalLink(link) | MdLink::WikiLink(link) => link.title.as_deref(),
        }
    }

    pub fn is_wikilink(&self) -> bool {
        match self {
            MdLink::NormalLink(_) => false,
            MdLink::WikiLink(_) => true,
        }
    }
}

#[derive(Debug)]
pub enum ResolvedLink<'a> {
    File {
        link: MdLink<'a>,
        /// The Uri of the file, the Link links to
        file_uri: &'a Url,
    },
    InternalHeading {
        link: MdLink<'a>,
        /// The Uri of the file, the Link links to
        file_uri: &'a Url,
        /// The Heading, the Link links to
        heading: &'a Heading,
    },
    ExternalHeading {
        link: MdLink<'a>,
        /// The Uri of the file, the Link links to
        file_uri: &'a Url,
        /// The Heading, the Link links to
        heading: &'a Heading,
    },
    Http,
    Unresolved,
}

impl ResolvedLink<'_> {
    /// The position of the Link
    pub fn link_position(&self) -> Option<&AstPosition> {
        match self {
            ResolvedLink::File { link, .. }
            | ResolvedLink::InternalHeading { link, .. }
            | ResolvedLink::ExternalHeading { link, .. } => link.position(),
            _ => None,
        }
    }

    /// The Url of the file, the Link links to
    pub fn file_uri(&self) -> Option<&Url> {
        match self {
            ResolvedLink::File { file_uri, .. }
            | ResolvedLink::InternalHeading { file_uri, .. }
            | ResolvedLink::ExternalHeading { file_uri, .. } => Some(file_uri),
            _ => None,
        }
    }

    /// If Link links to a heading, return the text of this heading
    pub fn heading_text(&self) -> Option<&str> {
        match self {
            ResolvedLink::InternalHeading { heading, .. }
            | ResolvedLink::ExternalHeading { heading, .. } => get_heading_text(heading),
            _ => None,
        }
    }
}

pub fn resolve_link<'a>(link: &'a Link, state: &'a State) -> ResolvedLink<'a> {
    let md_link = MdLink::new(link);

    if link.url.starts_with("http") {
        return ResolvedLink::Http;
    }

    match link.url.split_once('#') {
        Some(("", _)) => {
            for (url, md_file) in state.md_files.iter() {
                if let Some(heading) = &md_file.ast.find_heading_for_link(link) {
                    return ResolvedLink::InternalHeading {
                        link: md_link,
                        file_uri: url,
                        heading,
                    };
                }
            }
            ResolvedLink::Unresolved
        }
        // link with referece to heading `...#...`
        Some((file_ref_text, heading_ref_text)) => {
            let file = match md_link {
                MdLink::NormalLink(_) => url_decode(file_ref_text),
                MdLink::WikiLink(_) => {
                    // allow both for wikilinks: with suffix and without
                    if file_ref_text.ends_with(".md") {
                        file_ref_text.into()
                    } else {
                        format!("{}.md", file_ref_text)
                    }
                }
            };

            for (url, relative_path) in state.get_file_list() {
                if relative_path == file {
                    // as we get url from state it must be in there
                    let ast = state.ast_for_uri(url).unwrap();
                    if let Some(heading) = ast.find_heading_for_link_identifier(heading_ref_text) {
                        return ResolvedLink::ExternalHeading {
                            link: md_link,
                            file_uri: url,
                            heading,
                        };
                    }
                    return ResolvedLink::File {
                        link: md_link,
                        file_uri: url,
                    };
                }
            }
            ResolvedLink::Unresolved
        }
        None => {
            let file = if link.url.ends_with(".md") {
                link.url.to_string()
            } else {
                format!("{}.md", link.url)
            };
            for (url, relative_path) in state.get_file_list() {
                if relative_path == file {
                    return ResolvedLink::File {
                        link: md_link,
                        file_uri: url,
                    };
                }
            }
            ResolvedLink::Unresolved
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
struct ExtractedWikiLink {
    content: String,
    start_position: usize,
    line_number: usize,
}

impl ExtractedWikiLink {
    fn link_text_node(&self, start_line: usize) -> Node {
        let value = "".to_string(); // TODO parse Wikilinks with `|` (everything after is value)
        let value_len = value.len();
        let link_text = Text {
            value,
            position: Some(AstPosition {
                start: Point {
                    line: start_line + self.line_number,
                    column: self.start_position,
                    offset: self.start_position,
                },
                end: Point {
                    line: start_line + self.line_number,
                    column: self.start_position + value_len,
                    offset: self.start_position + value_len,
                },
            }),
        };
        Node::Text(link_text)
    }

    fn link_node(&self, start_line: usize) -> Node {
        let link_text_node = self.link_text_node(start_line);

        let link = Link {
            children: vec![link_text_node],
            position: Some(AstPosition {
                start: Point {
                    line: start_line + self.line_number,
                    column: self.start_position - 2,
                    offset: self.start_position - 2,
                },
                end: Point {
                    line: start_line + self.line_number,
                    column: self.start_position + self.content.len() + 2,
                    offset: self.start_position + self.content.len() + 2,
                },
            }),
            url: self.content.clone(),
            title: Some("wikilink".to_string()),
        };
        Node::Link(link)
    }
}

fn extract_wiki_links(input: &str) -> Vec<ExtractedWikiLink> {
    let re = Regex::new(r"\[\[([\s\S]*?)\]\]").unwrap();

    input
        .lines()
        .enumerate()
        .flat_map(|(line_number, line)| {
            re.captures_iter(line).filter_map(move |captures| {
                captures.get(1).map(|content| ExtractedWikiLink {
                    content: content.as_str().to_string(),
                    start_position: content.start() + 1,
                    line_number,
                })
            })
        })
        .collect()
}

pub fn parse_wiki_links(node: &mut Node) {
    let mut links = Vec::new();
    if let Some(children) = node.children() {
        for child in children {
            if let Node::Text(t) = child {
                if t.value.contains("[[") && t.value.contains("]]") {
                    let t_position = t.position.as_ref().unwrap().clone();
                    let extracted = extract_wiki_links(&t.value);

                    for i in extracted {
                        let link_ast = i.link_node(t_position.start.line);
                        links.push(link_ast);
                    }
                }
            }
        }
    }
    if let Some(children) = node.children_mut() {
        children.append(&mut links);
    }
    // recurse through children
    if let Some(children) = node.children_mut() {
        for child in children {
            parse_wiki_links(child);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_wiki_links() {
        let link_content_1 = "link content 1".to_string();
        let link_content_2 = "link content 2".to_string();
        let input = format!(
            "[[{}]]\n\n# Heading \n\n[[{}]]",
            link_content_1, link_content_2
        );
        let extracted = extract_wiki_links(&input);
        let expected_1 = ExtractedWikiLink {
            content: link_content_1,
            start_position: 3,
            line_number: 0,
        };
        let expected_2 = ExtractedWikiLink {
            content: link_content_2,
            start_position: 3,
            line_number: 4,
        };
        assert_eq!(extracted[0], expected_1);
        assert_eq!(extracted[1], expected_2);
    }
}
