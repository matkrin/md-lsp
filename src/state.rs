use std::{
    collections::HashMap,
    fs,
    path::{Path, PathBuf},
};

use ignore::Walk;
use itertools::Itertools;
use lsp_types::{Range, Url, WorkspaceFolder};
use markdown::mdast::Node;

use crate::links::parse_wiki_links;

#[derive(Debug)]
pub struct MdFile {
    buffer: String,
    pub ast: Node,
}

#[derive(Debug, Default)]
pub struct State {
    pub md_files: HashMap<Url, MdFile>,
    workspace_folder: Option<WorkspaceFolder>,
}

impl State {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn workspace_folder(&self) -> Option<&WorkspaceFolder> {
        self.workspace_folder.as_ref()
    }

    pub fn set_workspace_folder(&mut self, workspace_folder: WorkspaceFolder) {
        self.workspace_folder = Some(workspace_folder);
    }

    pub fn set_buffer(&mut self, uri: &Url, buffer: String) {
        let ast = markdown::to_mdast(&buffer, &markdown::ParseOptions::gfm()).unwrap();
        if let Some(md_file) = self.md_files.get_mut(uri) {
            md_file.ast = ast;
            parse_wiki_links(&mut md_file.ast);
            md_file.buffer = buffer;
        } else {
            let md_file = MdFile { buffer, ast };
            self.md_files.insert(uri.clone(), md_file);
        }
    }

    pub fn ast_for_uri(&self, uri: &Url) -> Option<&Node> {
        self.md_files.get(uri).map(|md_file| &md_file.ast)
    }

    pub fn buffer_for_uri(&self, uri: &Url) -> Option<&str> {
        self.md_files
            .get(uri)
            .map(|md_file| md_file.buffer.as_str())
    }

    pub fn index_md_files(&mut self, workspace_folders: &[WorkspaceFolder]) {
        let md_files = self.find_md_files(workspace_folders);
        self.md_files = md_files
            .into_iter()
            .map(|file| {
                log::info!("INDEXING: {:#?}", &file);
                let buffer = fs::read_to_string(&file).unwrap();
                let mut ast = markdown::to_mdast(&buffer, &markdown::ParseOptions::gfm()).unwrap();
                parse_wiki_links(&mut ast);
                let uri = Url::from_file_path(&file).unwrap();

                (uri, MdFile { buffer, ast })
            })
            .collect();
    }

    fn find_md_files(&self, workspace_folders: &[WorkspaceFolder]) -> Vec<PathBuf> {
        let mut md_files = Vec::new();
        for folder in workspace_folders {
            if let Ok(f) = folder.uri.to_file_path() {
                for entry in Walk::new(f).flatten() {
                    let path = entry.into_path();
                    if path.extension().is_some_and(|ext| ext == "md") {
                        md_files.push(path)
                    }
                }
            }
        }
        md_files
    }

    pub fn buffer_range_for_uri(&self, uri: &Url, range: &Range) -> Option<String> {
        let doc = self.buffer_for_uri(uri)?;
        let start_line = range.start.line as usize;
        let end_line = range.end.line as usize;
        let start_column = range.start.character as usize;
        let end_column = range.end.character as usize;

        let sliced = doc
            .lines()
            .enumerate()
            .filter_map(|(line_num, line)| {
                if line_num == start_line {
                    line.get(start_column..)
                } else if line_num == end_line {
                    line.get(..end_column)
                } else if line_num >= start_line && line_num <= end_line {
                    Some(line)
                } else {
                    None
                }
            })
            .join("\n");
        Some(sliced)
    }

    pub fn cursor_char(&self, uri: &Url, pos: &lsp_types::Position) -> Option<char> {
        let doc = self.buffer_for_uri(uri)?;
        let line = doc.lines().nth(pos.line as usize)?;
        line.chars().nth((pos.character.checked_sub(1))? as usize)
    }

    pub fn peek_behind_position(&self, uri: &Url, pos: &lsp_types::Position) -> Option<char> {
        let doc = self.buffer_for_uri(uri)?;
        let line = doc.lines().nth(pos.line as usize)?;
        line.chars().nth((pos.character.checked_sub(2))? as usize)
    }

    pub fn get_file_list(&self) -> Vec<(&Url, String)> {
        self.md_files
            .keys()
            .filter_map(|url| {
                self.workspace_folder().and_then(|wsf| {
                    let root = PathBuf::from(&wsf.uri.path());
                    let file_path = url.to_file_path().ok()?;
                    let path_from_root = path_from_root(&root, &file_path)?;
                    Some((url, path_from_root))
                })
            })
            .collect()
    }
}

pub fn path_from_root(from: &Path, to: &Path) -> Option<String> {
    if let Ok(rel) = to.strip_prefix(from) {
        Some(format!("/{}", rel.to_string_lossy()))
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn create_workspacefolder() -> WorkspaceFolder {
        let workspacefolder_dir = std::env::current_dir().expect("Current directory should exist");
        let s = format!("file:///{}", workspacefolder_dir.to_str().unwrap());
        let wsf_url = Url::parse(&s).expect("Parsing URL failed");
        let wsf_name = workspacefolder_dir.file_name().unwrap().to_str().unwrap();
        WorkspaceFolder {
            uri: wsf_url,
            name: wsf_name.into(),
        }
    }

    fn init_state() -> State {
        let workspace_folder = create_workspacefolder();
        let mut state = State::new();
        state.set_workspace_folder(workspace_folder);
        state
    }

    #[test]
    fn test_path_from_root() {
        let path1 = PathBuf::from("/foo/bar");
        let path2 = PathBuf::from("/foo/bar/baz");
        let path3 = PathBuf::from("/other/bar/baz");
        let from_root1 = path_from_root(&path1, &path2);
        assert_eq!(from_root1, Some("/baz".to_string()));
        let from_root2 = path_from_root(&path1, &path3);
        assert_eq!(from_root2, None);
    }

    #[test]
    fn test_index_md_files() {
        let workspace_folder = create_workspacefolder();
        let root_path = workspace_folder
            .uri
            .to_file_path()
            .expect("workspace_folder URI should be a valid file path");
        let mut state = init_state();
        state.index_md_files(&[workspace_folder]);
        assert!(!state.md_files.is_empty());
        let mut file_names: Vec<String> = state
            .md_files
            .keys()
            .filter_map(|url| {
                url.to_file_path()
                    .ok()
                    .and_then(|path| path.strip_prefix(&root_path).ok().map(|p| p.to_path_buf()))
                    .map(|rel_path| rel_path.display().to_string())
            })
            .collect();
        file_names.sort();
        insta::assert_debug_snapshot!(file_names);
    }
}
