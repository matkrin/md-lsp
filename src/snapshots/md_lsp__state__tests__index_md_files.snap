---
source: src/state.rs
expression: file_names
---
[
    "README.md",
    "testdata/test file c.md",
    "testdata/test-file-b.md",
    "testdata/test_file_a.md",
    "testdata/testfiled.md",
    "todo.md",
]
